### PR TITLE
Add similarity scores and metadata to Source Nodes

### PR DESCRIPTION
**Pull request by emptycrown**
Some improvements to what info is provided in the user-facing `SourceNodes`:
- Similarity score is now included, addressing #243 
- Metadata is now included (so you can immediately display info about the original document in the UI)
- Added 2 other similarity modes in addition to cosine similarity (dot product, euclidean distance)
- Added a QA template that allows the model to use prior knowledge

Included a test case for similarity score in vector scores

### CODE REVIEW

**Comment by emptycrown** [comment]
Included one more enhancement:
- New web reader using BeautifulSoup
- Allows for more fine-grained scraping of specific sites. Included an extractor for Substack posts

File: gpt_index/data_structs/data_structs.py [hunk 1]
```diff
@@ -51,6 +51,9 @@ def __post_init__(self) -> None:
     # reference document id
     ref_doc_id: Optional[str] = None
 
+    # extra metadata passed down from document
+    extra_info: Optional[Dict[str, Any]] = None
```
Comment (jerryjliu) [inline gpt_index/data_structs/data_structs.py:55]: don't need this, Node should already inherit this from BaseDocument

File: gpt_index/embeddings/base.py [hunk 2]
```diff
@@ -42,10 +51,21 @@ def get_text_embedding(self, text: str) -> List[float]:
...
 
-    def similarity(self, embedding1: EMB_TYPE, embedding2: EMB_TYPE) -> float:
+    def similarity(
+        self,
+        embedding1: EMB_TYPE,
+        embedding2: EMB_TYPE,
+        mode: SimilarityMode = SimilarityMode.DEFAULT,
+    ) -> float:
         """Get embedding similarity."""
+        if mode == SimilarityMode.EUCLIDEAN:
```
Comment (jerryjliu) [inline gpt_index/embeddings/base.py:61]: nit: can we do more explicit if/else if conditions?

e.g. 
```
if mode == SimilarityMode.EUCLIDEAN:
   ...
elif mode == SimilarityMode.DOT_PRODUCT:
   ...
else:
   ...
```

File: gpt_index/indices/query/vector_store/faiss.py [hunk 3]
```diff
@@ -53,12 +55,14 @@ def __init__(
         self._faiss_index = faiss_index
 
     def _get_nodes_for_response(
-        self, query_str: str, verbose: bool = False
+        self, query_str: str, verbose: bool = False, **nodes_kwargs: Any
```
Comment (jerryjliu) [inline gpt_index/indices/query/vector_store/faiss.py:58]: let's be explicit about keyword arguments (if we need similarity_tracker, which i'm not sure if we need yet).

File: gpt_index/indices/query/embedding_utils.py [hunk 4]
```diff
@@ -31,3 +32,32 @@ def get_top_k_embeddings(
     result_ids = [n for _, n in result_tups]
 
     return result_similarities, result_ids
+
+
+class SimilarityTracker:
+    """Helper class to manage node similarities during lifecycle of a single query."""
```
Comment (jerryjliu) [inline gpt_index/indices/query/embedding_utils.py:38]: Is the goal of this class mostly just to track similarities? 

I have an alternative proposal - there's already an `extra_info` field in Node. If similarity can be computed, then just update extra_info with `{"similarity": value}`. This way:
- You avoid needing to specify node_kwargs in _get_nodes_for_response
- You don't need to define a SimilarityTracker class

File: gpt_index/prompts/default_prompts.py [hunk 5]
```diff
@@ -108,6 +108,19 @@
...
 
+DEFAULT_TEXT_QA_PROMPT_W_KNOWLEDGE_TMPL = (
+    "Context information is below. \n"
+    "---------------------\n"
+    "{context_str}"
+    "\n---------------------\n"
+    "Given the context information and all your prior knowledge, "
+    "answer the question: {query_str}\n"
+)
+DEFAULT_TEXT_QA_W_KNOWLEDGE_PROMPT = QuestionAnswerPrompt(
```
Comment (jerryjliu) [inline gpt_index/prompts/default_prompts.py:119]: where is this used? can this be a separate PR?

**Review by jerryjliu — COMMENTED** [review]
Thanks for the changes! Some comments below. Also high-level note, there's really like 2-3 subchanges in here 🙂 - next time would be good to split it up a bit. 

> Metadata is now included (so you can immediately display info about the original document in the UI)

Where is this change?

File: gpt_index/data_structs/data_structs.py (hunk 1 above)
Comment (emptycrown) [reply gpt_index/data_structs/data_structs.py:55]: Hm yeah don't recall why I added that

**Comment by emptycrown** [comment]
> Thanks for the changes! Some comments below. Also high-level note, there's really like 2-3 subchanges in here 🙂 - next time would be good to split it up a bit.
> 
> > Metadata is now included (so you can immediately display info about the original document in the UI)
> 
> Where is this change?

`from_node` method of `SourceNode` now saves `extra_info` into the SourceNode

re: multiple PRs, for sure! Got a bit lazy here.

File: gpt_index/response/schema.py [hunk 6]
```diff
@@ -19,11 +19,20 @@ class SourceNode(DataClassJsonMixin):
 
     source_text: str
     doc_id: Optional[str]
+    extra_info: Optional[Dict[str, Any]] = None
```
Comment (jerryjliu) [inline gpt_index/response/schema.py:22]: conceptually extra_info is carried over from the original Document right? if that's the case do we need to store on SourceNode vs. just using doc_id to look up the original document?

Comment (jerryjliu) [reply gpt_index/response/schema.py:22]: i could see an argument for this if extra_info has more stuff (say from the Node) than just what's carried over from the original Document

File: gpt_index/indices/query/embedding_utils.py (hunk 4 above)
Comment (emptycrown) [reply gpt_index/indices/query/embedding_utils.py:38]: I briefly considered that, and would love to avoid the kwargs stuff, but isn't it kind of weird to intermingle similarity with metadata about the document that's passed down to the node? 

Moreover, the similarity score is specific to each *query* and should change each time, whereas my understanding is that the nodes exist at a higher level and their contents (including extra_info) don't change.

Comment (jerryjliu) [reply gpt_index/indices/query/embedding_utils.py:38]: Yeah you're right. What are your thoughts on the updated proposals?

File: gpt_index/response/schema.py (hunk 6 above)
Comment (emptycrown) [reply gpt_index/response/schema.py:22]: Yeah technically you can always look up the doc_id but I found it to be inconvenient because you'd have to perform the lookup for every SourceNode before returning the result to the Front End, since there's no concept of documents on the FE. Also, it's possible that in the future, a Node's extra_info may diverge from the Document's extra_info (like in your earlier proposal). To me, it feels natural to have the important attributes of the Node included in the Source Node.

IMO dream scenario is that you get a response from `index.query`, forward that to the FE, and you have everything you need to make a complete UX (including metadata and similarity scores).

File: gpt_index/indices/query/embedding_utils.py (hunk 4 above)
Comment (emptycrown) [reply gpt_index/indices/query/embedding_utils.py:38]: I'm good with (1). Actually, (2) was the first thing I tried but it felt really ugly especially since most of the child implementations of _get_nodes_for_response don't actually have similarities.

Comment (emptycrown) [reply gpt_index/indices/query/embedding_utils.py:38]: ^ edited

File: gpt_index/response/schema.py (hunk 6 above)
Comment (jerryjliu) [reply gpt_index/response/schema.py:22]: Hm ok! we can keep if it's easier for this given use case

File: gpt_index/indices/query/vector_store/faiss.py (hunk 3 above)
Comment (emptycrown) [reply gpt_index/indices/query/vector_store/faiss.py]: Done.

Comment (emptycrown) [reply gpt_index/indices/query/vector_store/faiss.py]: Probably misunderstood how kwargs were being used in the codebase. Makes sense now.

File: gpt_index/embeddings/base.py [hunk 7]
```diff
@@ -42,11 +51,22 @@ def get_text_embedding(self, text: str) -> List[float]:
...
+        embedding2: EMB_TYPE,
+        mode: SimilarityMode = SimilarityMode.DEFAULT,
+    ) -> float:
         """Get embedding similarity."""
-        product = np.dot(embedding1, embedding2)
-        norm = np.linalg.norm(embedding1) * np.linalg.norm(embedding2)
-        return product / norm
+        if mode == SimilarityMode.EUCLIDEAN:
+            return float(np.linalg.norm(np.array(embedding1) - np.array(embedding2)))
+        elif mode == SimilarityMode.DOT_PRODUCT:
```
Comment (jerryjliu) [inline gpt_index/embeddings/base.py:63]: might be good to add a test for similarity modes

**Review by jerryjliu — APPROVED** [review]
